pingcap/tidb-operator
# support run tiflash pod with ConfigMap mounted in tiflash container

### What problem does this PR solve?

Add an annotation `tiflash.tidb.pingcap.com/mount-cm-in-tiflash-container` to tc.spec.tiflash.annotations to supports run tiflash pod without init container and directly mount ConfigMap in tiflash container.  This change enable tiflash supports config hot-reload when the tc.spec.configUpdateStrategy is set to InPlace.

### What is changed and how does it work?
When this annotation is annotated, the tiflash container will directly mount config ConfigMap and use new start script to launch tiflash without the initContainer did ever that use sed to modify ConfigMap and then generate a new config files into /data0 directory. 


### Code changes


### Tests



#### Test steps
* create a TC with tiflash enabled
* edit TC to add `tiflash.tidb.pingcap.com/mount-cm-in-tiflash-container: "true"` into tc.spec.tiflash.annotations
* old tiflash pod will be terminated and new tiflash pod without initContainer, will be created
* modify the ConfigMap used by the tiflash pod to change `config_templ.toml:logger.level` to "debug" and add `config_templ.toml:logger.count = 11`
* watch log of tiflash pod, some logs like 'Loading config ' should be logged

#### Diffs of key resources
* the diffs of tiflash pod 
```diff
diff old.txt new.txt -I 'kube-api-access'
13c13,26
<     - /tiflash/tiflash server --config-file /data0/config.toml
---
>     - |
>       #!/bin/sh
>
>       # This script is used to start tiflash containers in kubernetes cluster
>
>       # Use DownwardAPIVolumeFiles to store informations of the cluster:
>       # https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#the-downward-api
>       #
>       #   runmode="normal/debug"
>       #
>
>       set -uo pipefail
>
>       ANNOTATIONS="/etc/podinfo/annotations"
>
>       if [[ ! -f "${ANNOTATIONS}" ]]
>       then
>           echo "${ANNOTATIONS} does't exist, exiting."
>           exit 1
>       fi
>       source ${ANNOTATIONS} 2>/dev/null
>
>       runmode=${runmode:-normal}
>       if [[ X${runmode} == Xdebug ]]
>       then
>           echo "entering debug mode."
>           tail -f /dev/null
>       fi
>
>       # Use HOSTNAME if POD_NAME is unset for backward compatibility.
>       POD_NAME=${POD_NAME:-$HOSTNAME}
>       PD_ADDRESS="https://${CLUSTER_NAME}-pd:2379"
>
>       ARGS="server --config-file /etc/tiflash/config_templ.toml \
>       -- \
>       --flash.proxy.advertise-addr=${POD_NAME}.${HEADLESS_SERVICE_NAME}.${NAMESPACE}.svc:20170 \
>       --flash.proxy.advertise-status-addr=${POD_NAME}.${HEADLESS_SERVICE_NAME}.${NAMESPACE}.svc:20292 \
>       --flash.service_addr=0.0.0.0:3930 \
>       --raft.pd_addr=${PD_ADDRESS}
>       "
>
>       if [ ! -z "${STORE_LABELS:-}" ]; then
>         LABELS=" --labels ${STORE_LABELS} "
>         ARGS="${ARGS}${LABELS}"
>       fi
>
>       echo "starting tiflash-server ..."
>       echo "/tiflash/tiflash ${ARGS}"
>       exec /tiflash/tiflash ${ARGS}
79a128,130
>     - mountPath: /etc/tiflash
>       name: config
>       readOnly: true
153,181c204
<   initContainers:
<   - command:
<     - sh
<     - -c
<     - set -ex;ordinal=`echo ${POD_NAME} | awk -F- '{print $NF}'`;sed s/POD_NUM/${ordinal}/g
<       /etc/tiflash/config_templ.toml > /data0/config.toml;sed s/POD_NUM/${ordinal}/g
<       /etc/tiflash/proxy_templ.toml > /data0/proxy.toml
<     env:
<     - name: POD_NAME
<       valueFrom:
<         fieldRef:
<           apiVersion: v1
<           fieldPath: metadata.name
<     image: 385595570414.dkr.ecr.us-west-2.amazonaws.com/third-party/busybox:1.34.1
<     imagePullPolicy: IfNotPresent
<     name: init
<     resources: {}
<     terminationMessagePath: /dev/termination-log
<     terminationMessagePolicy: File
<     volumeMounts:
<     - mountPath: /data0
<       name: data0
<     - mountPath: /etc/tiflash
<       name: config
<       readOnly: true
<     - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
<       name: kube-api-access-dwb6m
<       readOnly: true
```
* the diffs of ConfigMap
```diff
<         advertise-addr = "db-tiflash-POD_NUM.db-tiflash-peer.tidb9527.svc:20170"
<         config = "/data0/proxy.toml"
---
>         config = "/etc/proxy_templ.toml"
55,56d53
<       advertise-status-addr = "db-tiflash-POD_NUM.db-tiflash-peer.tidb9527.svc:20292"
<       engine-addr = "db-tiflash-POD_NUM.db-tiflash-peer.tidb9527.svc:3930"
```
* diffs between command line of tiflash container
```diff
<   /tiflash/tiflash server --config-file /data0/config.toml
---
>   /tiflash/tiflash server --config-file /etc/tiflash/config_templ.toml -- --flash.proxy.advertise-addr=db-tiflash-0.db-tiflash-peer.tidb9527.svc:20170 --flash.proxy.advertise-status-addr=db-tiflash-0.db-tiflash-peer.tidb9527.svc:20292 --flash.service_addr=0.0.0.0:3930 --raft.pd_addr=https://db-pd:2379
```
### Side effects


### Related changes


### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```
